packages/panzer/expr-eval/test/ExprEvalTest.cpp

```diff
@@ -0,0 +1,220 @@
+// @HEADER
+// ***********************************************************************
+//
+//           Panzer: A partial differential equation assembly
+//       engine for strongly coupled complex multiphysics systems
+//                 Copyright (2011) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Roger P. Pawlowski (rppawlo@sandia.gov) and
+// Eric C. Cyr (eccyr@sandia.gov)
+// ***********************************************************************
+// @HEADER
+
+#include <Teuchos_UnitTestHarness.hpp>
+
+#include "Panzer_ExprEval_impl.hpp"
+
+namespace panzer {
+
+TEUCHOS_UNIT_TEST(ExprEval, test_1d_system)
+{
+  Expr::Eval<double*> eval;
+  Expr::set_cmath_functions(eval);
+  Teuchos::any result;
+  eval.read_string(result, "1+1", "one plus one");
+  Kokkos::View<double const> x = Teuchos::any_cast<Kokkos::View<double const>>(result);
```

Comment:
How long does this thing persist? Will we have to do this case all the time, or is it just once?

Also, can we launch this execution from the device instead.

Reply:
> How long does this thing persist? Will we have to do this case all the time, or is it just once?

If you mean the result, currently expression results are allocated at each operator, so it is created when the `+` is evaluated and persists only as long as that `Teuchos::any` is in use and not overwritten by another evaluation.

> Also, can we launch this execution from the device instead.

No, the `Teuchos::Parser` part won't run on the device. The idea is that to use the device, you collect all the input values for a particular variable into a single `Kokkos::View<..., Kokkos::Cuda, ...>`, and run this once on the host, which dispatches work to the device.

Reply:
Okay. So each time this is executed the string is parsed again? Is it possible to encode this in some other "parsed" object that does the execution?

Reply:
@eric-c-cyr that was the approach taken by other systems that do this, where you would call this at every evaluation point with some kind of pre-compiled byte code. The approach taken here is fundamentally different in that each value should be a large view containing data for all the evaluation points, so this should only be called once per side set for example. That amortizes the overhead of the parsing and should remove the need for any kind of parsing optimization.

Reply:
I understand this point.  But what will happen is this will get reevaluated many times (over all the quadrature points).